src/core/resource.cc

```diff
@@ -619,28 +619,44 @@ resources allocate(configuration& c) {
     std::vector<std::pair<cpu, size_t>> remains;
 
     auto cpu_sets = distribute_objects(topology, procs);
-    auto num_nodes = hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_NUMANODE);
 
     for (auto&& cs : cpu_sets()) {
         auto [cpu_id, node] = get_cpu_id_and_numa_node_for_set(topology, cs);
-        if (node == nullptr) {
+        if (node == nullptr || !get_memory_from_hwloc_obj(node)) {
             orphan_pus.push_back(cpu_id);
         } else {
```

Comment:
I don't understand this special case. What's the difference between a node that has a small amount of memory and no memory? Should work exactly the same.

Maybe you're fast-promoting it to only participate in the second round of distribution.

Reply:
So what if you run with --smp 1 and that happens to be on a memory-less NUMA node?

It will trigger the fallback even though hwloc works normally.

I think the algorithm would work if we referenced all numa nodes, not just those in cpu_to_node.

```
collection\<numa_node> numas;
map\<cpu, numa_node> assignment;

for (all cpus) {
    try to assign from assignment[cpu]
}

for (all cpus) {
     delta = whatever we failed to assign in phase 1
     assign delta from whatever node as free memory
}
```